### ISSUE TITLE
Make VaultSharp fail fast on Login Authentication

### ISSUE THREAD

**Issue by rajanadar** [rajanadar/VaultSharp#177]
> I'll check with @intechdebt but I dont see any issue with the vaultsharp code. Ideally it would be helpful if the code detected the kerberos authenticaion failure. We only get an error when we make the call to say read a secret. And the error says the action is unauthorised.

That's a good suggestion @purplemog . One of the challenges is when to attempt the authentication. Currently in VaultSharp, it does a deferred authentication when you attempt to perform the first real operation. (and not when you initialize the Auth info) My rationale for it, was to ensure you get a vault token that is as much alive as possible. However I do the downsides of it. No fast fail. Too late when the error happens.

Fast failure is preferred when there are auth failures. So let me add a flag to VaultClient settings (PerformImmediateAuth) which will attempt to login immediately when you are initializing VaultClient. And it should fail fast.

_Originally posted by @rajanadar in https://github.com/rajanadar/VaultSharp/issues/170#issuecomment-766398572_

**Comment by rajanadar** [comment]
Fail fast. And try to see if VaultSharp can emit error messages from Vault API coming in error or even warning fields.

**Comment by intechdebt** [comment]
Thanks @rajanadar 

We'll keep an eye for your release with the fast fail feature and will give it go.

By the way.... We are still awaiting for the SPN change to be implemented before we can retest the Kerberos authentication.

**Comment by rajanadar** [comment]
https://www.nuget.org/packages/VaultSharp/1.6.0.3
https://github.com/rajanadar/VaultSharp/blob/master/README.md#vaultsharp-and-immediate-login-failure-detection
